Army-Java/Units/Unit.java

```diff
@@ -0,0 +1,245 @@
+package com.gymfox.Army.Units;
+
+import com.gymfox.Army.Ability.Ability;
+import com.gymfox.Army.Exception.IsSelfAttackException;
+import com.gymfox.Army.Exception.MasterAttackedException;
+import com.gymfox.Army.Exception.UnitIsDeadException;
+import com.gymfox.Army.Observer.Observable;
+import com.gymfox.Army.Observer.Observer;
+import com.gymfox.Army.State.State;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+
+public abstract class Unit implements Observable, Observer {
+
+    public enum UnitType {
+        SOLDIER,
+        ROGUE,
+        BERSERK,
+        VAMPIRE,
+        WEREWOLF,
+        WIZARD,
+        HEALER,
+        PRIEST,
+        WARLOCK,
+        NECROMANCER
+    }
+
+    private String name;
+    private int healthPointLimit;
+    private int currentHP;
+    private int damage;
+    private boolean isDead = false;
+    private UnitType unitType;
+    private Ability ability;
+    private State currentState;
+    private State nextState;
```

Comment:
I'd move States to abilities. That should make states usage more flexible.
In default ability can throw some `SingleStateException` when `changeState()` invoked. In abilities that should have abilities (werewolf) should just override this method and handle all logic